File: android/src/main/java/com/befovy/fijkplayer/FijkPlugin.java
//MIT License
//
//Copyright (c) [2019-2020] [Befovy]
//
//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:
//
//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.
//
//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

package com.befovy.fijkplayer;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.content.res.Configuration;
import android.media.AudioAttributes;
import android.media.AudioFocusRequest;
import android.media.AudioManager;
import android.os.Build;
import android.provider.Settings;
import android.text.TextUtils;
import android.util.Log;
import android.util.SparseArray;
import android.view.KeyEvent;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.flutter.view.TextureRegistry;
import tv.danmaku.ijk.media.player.IjkMediaPlayer;

/**
 * FijkPlugin
 */
public class FijkPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware, FijkEngine, FijkVolume.VolumeKeyListener, AudioManager.OnAudioFocusChangeListener {

    // show system volume changed UI if no playable player
    // hide system volume changed UI if some players are in playable state
    private static final int NO_UI_IF_PLAYABLE = 0;
    // show system volume changed UI if no start state player
    // hide system volume changed UI if some players are in start state
    private static final int NO_UI_IF_PLAYING = 1;
    // never show system volume changed UI
    @SuppressWarnings("unused")
    private static final int NEVER_SHOW_UI = 2;
    // always show system volume changed UI
    private static final int ALWAYS_SHOW_UI = 3;

    final private SparseArray<FijkPlayer> fijkPlayers = new SparseArray<>();

    private final QueuingEventSink mEventSink = new QueuingEventSink();

    private WeakReference<Activity> mActivity;
    private WeakReference<Context> mContext;
    private Registrar mRegistrar;
    private FlutterPluginBinding mBinding;

    // Count of playable players
    private int playableCnt = 0;
    // Count of playing players
    private int playingCnt = 0;
    private int volumeUIMode = ALWAYS_SHOW_UI;
    private float volStep = 1.0f / 16.0f;
    private boolean eventListening = false;
    // non-local field prevent GC
    private EventChannel mEventChannel;
    private Object mAudioFocusRequest;
    private boolean mAudioFocusRequested = false;


    /**
     * Plugin registration.
     */
    @SuppressWarnings("unused")
    public static void registerWith(Registrar registrar) {
        final MethodChannel channel = new MethodChannel(registrar.messenger(), "befovy.com/fijk");
        FijkPlugin plugin = new FijkPlugin();
        plugin.initWithRegistrar(registrar);
        channel.setMethodCallHandler(plugin);

        final FijkPlayer player = new FijkPlayer(plugin, true);
        player.setupSurface();
        player.release();
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
        final MethodChannel channel = new MethodChannel(binding.getBinaryMessenger(), "befovy.com/fijk");
        initWithBinding(binding);
        channel.setMethodCallHandler(this);

        final FijkPlayer player = new FijkPlayer(this, true);
        player.setupSurface();
        player.release();

        AudioManager audioManager = audioManager();
        if (audioManager != null) {
            int max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
            volStep = Math.max(1.0f / (float) max, volStep);
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        mContext = null;
    }

    @Override
    public void onAttachedToActivity(ActivityPluginBinding binding) {
        mActivity = new WeakReference<>(binding.getActivity());
        if (mActivity.get() instanceof FijkVolume.CanListenVolumeKey) {
            FijkVolume.CanListenVolumeKey canListenVolumeKey = (FijkVolume.CanListenVolumeKey) mActivity.get();
            canListenVolumeKey.setVolumeKeyListener(this);
        }
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        mActivity = null;
    }

    @Override
    public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
        mActivity = new WeakReference<>(binding.getActivity());
        if (mActivity.get() instanceof FijkVolume.CanListenVolumeKey) {
            FijkVolume.CanListenVolumeKey canListenVolumeKey = (FijkVolume.CanListenVolumeKey) mActivity.get();
            canListenVolumeKey.setVolumeKeyListener(this);
        }
    }

    @Override
    public void onDetachedFromActivity() {
        mActivity = null;
    }

    @Override
    @Nullable
    public TextureRegistry.SurfaceTextureEntry createSurfaceEntry() {
        if (mBinding != null) {
            return mBinding.getTextureRegistry().createSurfaceTexture();
        } else if (mRegistrar != null) {
            return mRegistrar.textures().createSurfaceTexture();
        }
        return null;
    }

    @Override
    @Nullable
    public BinaryMessenger messenger() {
        if (mBinding != null) {
            return mBinding.getBinaryMessenger();
        } else if (mRegistrar != null) {
            return mRegistrar.messenger();
        }
        return null;
    }

    @Override
    @Nullable
    public Context context() {
        if (mContext != null)
            return mContext.get();
        else
            return null;
    }

    @Nullable
    private Activity activity() {
        if (mRegistrar != null) {
            return mRegistrar.activity();
        } else if (mActivity != null) {
            return mActivity.get();
        } else {
            return null;
        }
    }

    @Override
    @Nullable
    public String lookupKeyForAsset(@NonNull String asset, @Nullable String packageName) {
        String path = null;
        if (mBinding != null) {
            if (TextUtils.isEmpty(packageName)) {
                path = mBinding.getFlutterAssets().getAssetFilePathByName(asset);
            } else {
                //noinspection ConstantConditions
                path = mBinding.getFlutterAssets().getAssetFilePathByName(asset, packageName);
            }
        } else if (mRegistrar != null) {
            if (TextUtils.isEmpty(packageName)) {
                path = mRegistrar.lookupKeyForAsset(asset);
            } else {
                path = mRegistrar.lookupKeyForAsset(asset, packageName);
            }
        }
        return path;
    }


    private void initWithRegistrar(@NonNull Registrar registrar) {
        mRegistrar = registrar;
        mContext = new WeakReference<>(registrar.activeContext());
        init(registrar.messenger());
    }

    private void initWithBinding(@NonNull FlutterPluginBinding binding) {
        mBinding = binding;
        mContext = new WeakReference<>(binding.getApplicationContext());
        init(binding.getBinaryMessenger());
    }

    /**
     * Maybe call init more than once
     *
     * @param messenger BinaryMessenger from flutter engine
     */
    private void init(BinaryMessenger messenger) {
        if (mEventChannel != null) {
            mEventChannel.setStreamHandler(null);
            mEventSink.setDelegate(null);
        }
        mEventChannel = new EventChannel(messenger, "befovy.com/fijk/event");
        mEventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mEventSink.setDelegate(eventSink);
            }

            @Override
            public void onCancel(Object o) {
                mEventSink.setDelegate(null);
            }
        });

        AudioManager audioManager = audioManager();
        if (audioManager != null) {
            int max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
            volStep = Math.max(1.0f / (float) max, volStep);
        }
    }


    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        Activity activity;
        switch (call.method) {
            case "getPlatformVersion":
                result.success("Android " + android.os.Build.VERSION.RELEASE);
                break;
            case "init":
                Log.i("FLUTTER", "call init:" + call.arguments.toString());
                result.success(null);
                break;
            case "createPlayer": {
                FijkPlayer fijkPlayer = new FijkPlayer(this, false);
                int playerId = fijkPlayer.getPlayerId();
                fijkPlayers.append(playerId, fijkPlayer);
                result.success(playerId);
                break;
            }
            case "releasePlayer": {
                int pid = -1;
                final Integer arg = call.argument("pid");
                if (arg != null)
                    pid = arg;
                FijkPlayer fijkPlayer = fijkPlayers.get(pid);
                if (fijkPlayer != null) {
                    fijkPlayer.release();
                    fijkPlayers.delete(pid);
                }
                result.success(null);
                break;
            }
            case "logLevel": {
                int level = 500;
                final Integer l = call.argument("level");
                if (l != null)
                    level = l;
                level = level / 100;
                level = Math.max(level, 0);
                level = Math.min(level, 8);
                IjkMediaPlayer.loadLibrariesOnce(null);
                IjkMediaPlayer.setLogLevel(level);
                result.success(null);
                break;
            }
            case "setOrientationPortrait":
                boolean changedPort = false;
                activity = activity();
                if (activity != null) {
                    int current_orientation = activity.getResources().getConfiguration().orientation;
                    if (current_orientation == Configuration.ORIENTATION_LANDSCAPE) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                            activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_USER_PORTRAIT);
                        } else {
                            activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR_PORTRAIT);
                        }
                        changedPort = true;
                    }
                }
                result.success(changedPort);
                break;
            case "setOrientationLandscape":
                boolean changedLand = false;
                activity = activity();
                if (activity != null) {
                    int current_orientation = activity.getResources().getConfiguration().orientation;
                    if (current_orientation == Configuration.ORIENTATION_PORTRAIT) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                            activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_USER_LANDSCAPE);
                        } else {
                            activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR_LANDSCAPE);
                        }
                        changedLand = true;
                    }
                }
                result.success(changedLand);
                break;
            case "setOrientationAuto":
                activity = activity();
                if (activity != null) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
                        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_USER);
                    } else {
                        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_FULL_SENSOR);
                    }
                }
                result.success(null);
                break;
            case "setScreenOn":
                boolean screenOn = false;
                if (call.hasArgument("on")) {
                    Boolean on = call.argument("on");
                    screenOn = on != null ? on : false;
                }
                setScreenOn(screenOn);
                result.success(null);
                break;
            case "isScreenKeptOn":
                result.success(isScreenKeptOn());
                break;
            case "brightness":
                float brightnessGot = getScreenBrightness();
                result.success(brightnessGot);
                break;
            case "setBrightness":
                if (call.hasArgument("brightness")) {
                    final Double var = call.argument("brightness");
                    if (var != null) {
                        float brightness = var.floatValue();
                        setScreenBrightness(brightness);
                    }
                }
                break;
            case "requestAudioFocus":
                audioFocus(true);
                result.success(null);
                break;
            case "releaseAudioFocus":
                audioFocus(false);
                result.success(null);
                break;
            case "volumeDown":
                float stepDown = volStep;
                if (call.hasArgument("step")) {
                    final Double step = call.argument("step");
                    stepDown = step != null ? step.floatValue() : stepDown;
                }
                result.success(volumeDown(stepDown));
                break;
            case "volumeUp":
                float stepUp = volStep;
                if (call.hasArgument("step")) {
                    final Double step = call.argument("step");
                    stepUp = step != null ? step.floatValue() : stepUp;
                }
                result.success(volumeUp(stepUp));
                break;
            case "volumeMute":
                result.success(volumeMute());
                break;
            case "systemVolume":
                result.success(systemVolume());
                break;
            case "volumeSet":
                float vol = systemVolume();
                final Double v = call.argument("vol");
                if (v != null) {
                    vol = setSystemVolume(v.floatValue());
                }
                result.success(vol);
                break;
            case "volUiMode":
                final Integer mode = call.argument("mode");
                if (mode != null) {
                    volumeUIMode = mode;
                }
                result.success(null);
                break;
            case "onLoad":
                eventListening = true;
                result.success(null);
                break;
            case "onUnload":
                eventListening = false;
                result.success(null);
                break;
            default:
                Log.w("FLUTTER", "onMethod Call, name: " + call.method);
                result.notImplemented();
                break;
        }
    }


    @Override
    public void onPlayingChange(int delta) {
        playingCnt += delta;
    }

    @Override
    public void onPlayableChange(int delta) {
        playableCnt += delta;
    }

    @Override
    public void onAudioFocusChange(int focusChange) {
        switch (focusChange) {
            case AudioManager.AUDIOFOCUS_GAIN:
                break;
            case AudioManager.AUDIOFOCUS_LOSS:
            case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                mAudioFocusRequested = false;
                mAudioFocusRequest = null;
                break;
        }
        Log.i("FIJKPLAYER", "onAudioFocusChange: " + focusChange);
    }

    /**
     * Set screen on enable or disable
     *
     * @param on true to set keep screen on enable
     *           false to set keep screen on disable
     */
    @Override
    public void setScreenOn(boolean on) {
        Activity activity = activity();
        if (activity == null || activity.getWindow() == null)
            return;
        if (on) {
            activity.getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        } else {
            activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }
    }

    /**
     * Check if screen is kept on
     *
     * @return true if screen is kept on
     */
    private boolean isScreenKeptOn() {
        Activity activity = activity();
        if (activity == null || activity.getWindow() == null)
            return false;
        int flag = activity.getWindow().getAttributes().flags;
        return (flag & WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) != 0;
    }


    private float getScreenBrightness() {
        Activity activity = activity();
        if (activity == null || activity.getWindow() == null)
            return 0;
        float brightness = activity.getWindow().getAttributes().screenBrightness;
        if (brightness < 0) {
            Context context = context();
            Log.w("FIJKPLAYER", "window attribute brightness less than 0");
            try {
                if (context != null) {
                    brightness = Settings.System.getInt(context.getContentResolver(), Settings.System.SCREEN_BRIGHTNESS) / (float) 255;
                }
            } catch (Settings.SettingNotFoundException e) {
                Log.e("FIJKPLAYER", "System brightness settings not found");
                brightness = 1.0f;
            }
        }
        return brightness;
    }

    private void setScreenBrightness(float brightness) {
        Activity activity = activity();
        if (activity == null || activity.getWindow() == null)
            return;
        WindowManager.LayoutParams layoutParams = activity.getWindow().getAttributes();
        layoutParams.screenBrightness = brightness;
        activity.getWindow().setAttributes(layoutParams);
    }

    @TargetApi(26)
    @SuppressWarnings("deprecation")
    private void requestAudioFocus() {
        AudioManager audioManager = audioManager();
        if (audioManager == null)
            return;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            AudioAttributes audioAttributes =
                    new AudioAttributes.Builder()
                            .setUsage(AudioAttributes.USAGE_MEDIA)
                            .setContentType(AudioAttributes.CONTENT_TYPE_MOVIE)
                            .build();

            AudioFocusRequest audioFocusRequest =
                    new AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
                            .setAudioAttributes(audioAttributes)
                            .setAcceptsDelayedFocusGain(true)
                            .setOnAudioFocusChangeListener(this) // Need to implement listener
                            .build();
            mAudioFocusRequest = audioFocusRequest;
            audioManager.requestAudioFocus(audioFocusRequest);
        } else {
            audioManager.requestAudioFocus(this,
                    AudioManager.STREAM_MUSIC, AudioManager.AUDIOFOCUS_GAIN);
        }
        mAudioFocusRequested = true;
    }

    @TargetApi(26)
    // @SuppressWarnings("deprecation")
    private void abandonAudioFocus() {
        AudioManager audioManager = audioManager();
        if (audioManager == null)
            return;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            if (mAudioFocusRequest != null) {
                audioManager.abandonAudioFocusRequest((AudioFocusRequest) mAudioFocusRequest);
                mAudioFocusRequest = null;
            }
        } else {
            audioManager.abandonAudioFocus(this);
        }
        mAudioFocusRequested = false;
    }

    /**
     * @param request true to request audio focus
     *                false to release audio focus
     */
    @Override
    public void audioFocus(boolean request) {
        Log.i("FIJKPLAYER", "audioFocus " + (request ? "request" : "release") + " state:" + mAudioFocusRequested);
        if (request && !mAudioFocusRequested) {
            requestAudioFocus();
        } else if (mAudioFocusRequested) {
            abandonAudioFocus();
        }
    }

    @Nullable
    private AudioManager audioManager() {
        Context context = context();
        if (context != null) {
            return (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        } else {
            Log.e("FIJKPLAYER", "context null, can't get AudioManager");
            return null;
        }
    }

    private float systemVolume() {
        AudioManager audioManager = audioManager();
        if (audioManager != null) {
            float max = (float) audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
            float vol = (float) audioManager.getStreamVolume(AudioManager.STREAM_MUSIC);
            return vol / max;
        } else {
            return 0.0f;
        }
    }

    private void sendVolumeEvent() {
        if (eventListening) {
            int flag = getVolumeChangeFlag();
            boolean showOsUI = (flag & AudioManager.FLAG_SHOW_UI) > 0;
            Map<String, Object> event = new HashMap<>();
            event.put("event", "volume");
            event.put("sui", showOsUI);
            event.put("vol", systemVolume());
            mEventSink.success(event);
        }
    }

    private float volumeUp(float step) {
        float vol = systemVolume();
        vol = vol + step;
        vol = setSystemVolume(vol);
        return vol;
    }

    private float volumeDown(float step) {
        float vol = systemVolume();
        vol = vol - step;
        vol = setSystemVolume(vol);
        return vol;
    }

    @SuppressWarnings("SameReturnValue")
    private float volumeMute() {
        setSystemVolume(0.0f);
        return 0.0f;
    }

    private int getVolumeChangeFlag() {
        int flag = 0;
        if (volumeUIMode == ALWAYS_SHOW_UI) {
            flag = AudioManager.FLAG_SHOW_UI;
        } else if (volumeUIMode == NO_UI_IF_PLAYING && playingCnt == 0) {
            flag = AudioManager.FLAG_SHOW_UI;
        } else if (volumeUIMode == NO_UI_IF_PLAYABLE && playableCnt == 0) {
            flag = AudioManager.FLAG_SHOW_UI;
        }
        return flag;
    }

    private float setSystemVolume(float vol) {
        int flag = getVolumeChangeFlag();
        AudioManager audioManager = audioManager();
        if (audioManager != null) {
            int max = audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
            int volIndex = (int) (vol * max);
            volIndex = Math.min(volIndex, max);
            volIndex = Math.max(volIndex, 0);
            audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, volIndex, flag);
            sendVolumeEvent();
            return (float) volIndex / (float) max;
        } else {
            return vol;
        }
    }

    @Override
    public boolean onVolumeKeyDown(int keyCode, KeyEvent event) {
        switch (keyCode) {
            case KeyEvent.KEYCODE_VOLUME_DOWN:
                volumeDown(volStep);
                return true;
            case KeyEvent.KEYCODE_VOLUME_UP:
                volumeUp(volStep);
                return true;
            case KeyEvent.KEYCODE_VOLUME_MUTE:
                volumeMute();
                return true;
        }
        return false;
    }
}

File: ios/Classes/FijkPlugin.m
// MIT License
//
// Copyright (c) [2019] [Befovy]
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#import "FijkPlugin.h"
#import "FijkPlayer.h"
#import "FijkQueuingEventSink.h"

#import <AVKit/AVKit.h>
#import <Flutter/Flutter.h>
#import <IJKMediaPlayer/IJKMediaPlayer.h>
#import <MediaPlayer/MediaPlayer.h>

typedef NS_ENUM(int, FijkVoUIMode) {
    hideUIWhenPlayable = 0,
    hideUIWhenPlaying,
    neverShowUI,
    alwaysShowUI,
};

@implementation FijkPlugin {
    NSObject<FlutterPluginRegistrar> *_registrar;
    NSMutableDictionary<NSNumber *, FijkPlayer *> *_fijkPlayers;

    FijkQueuingEventSink *_eventSink;
    FlutterEventChannel *_eventChannel;

    MPVolumeView *_volumeView;
    UISlider *_volumeViewSlider;
    BOOL _volumeInWindow;

    int _volumeUIMode;
    BOOL _eventListening;
    float _volStep;
    BOOL _showOsUI;
}

static FijkPlugin *_instance = nil;

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel =
        [FlutterMethodChannel methodChannelWithName:@"befovy.com/fijk"
                                    binaryMessenger:[registrar messenger]];
    FijkPlugin *instance = [[FijkPlugin alloc] initWithRegistrar:registrar];
    _instance = instance;
    [registrar addMethodCallDelegate:instance channel:channel];

    FijkPlayer *player = [[FijkPlayer alloc] initJustTexture];
    int64_t vid = [[registrar textures] registerTexture:player];
    [player shutdown];
    [[registrar textures] unregisterTexture:vid];
}

+ (FijkPlugin *)singleInstance {
    return _instance;
}

- (instancetype)initWithRegistrar:
    (NSObject<FlutterPluginRegistrar> *)registrar {
    self = [super init];
    if (self) {
        _registrar = registrar;
        _fijkPlayers = [[NSMutableDictionary alloc] init];
        _eventListening = FALSE;
        _volumeUIMode = alwaysShowUI;
        _volStep = 1.0 / 16.0;
        _showOsUI = YES;
        _eventSink = [[FijkQueuingEventSink alloc] init];

        _eventChannel =
            [FlutterEventChannel eventChannelWithName:@"befovy.com/fijk/event"
                                      binaryMessenger:[registrar messenger]];
        [_eventChannel setStreamHandler:self];

        [[NSNotificationCenter defaultCenter]
            addObserver:self
               selector:@selector(volumeChange:)
                   name:@"AVSystemController_SystemVolumeDidChangeNotification"
                 object:nil];
    }
    return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call
                  result:(FlutterResult)result {

    NSDictionary *argsMap = call.arguments;
    if ([@"getPlatformVersion" isEqualToString:call.method]) {
        NSString *osVersion = [[UIDevice currentDevice] systemVersion];
        result([@"iOS " stringByAppendingString:osVersion]);
    } else if ([@"init" isEqualToString:call.method]) {
        NSLog(@"FLUTTER: %s %@", "call init:", argsMap);
        result(NULL);
    } else if ([@"createPlayer" isEqualToString:call.method]) {
        FijkPlayer *fijkplayer =
            [[FijkPlayer alloc] initWithRegistrar:_registrar];
        NSNumber *playerId = fijkplayer.playerId;
        _fijkPlayers[playerId] = fijkplayer;
        result(playerId);
    } else if ([@"releasePlayer" isEqualToString:call.method]) {
        NSNumber *pid = argsMap[@"pid"];
        FijkPlayer *fijkPlayer = [_fijkPlayers objectForKey:pid];
        [fijkPlayer shutdown];
        if (fijkPlayer != nil) {
            [_fijkPlayers removeObjectForKey:pid];
        }
        result(nil);
    } else if ([@"logLevel" isEqualToString:call.method]) {
        NSNumber *level = argsMap[@"level"];
        int l = [level intValue] / 100;
        l = l < 0 ? 0 : l;
        l = l > 8 ? 8 : l;
        [IJKFFMoviePlayerController setLogLevel:l];
        result(nil);
    } else if ([@"setOrientationPortrait" isEqualToString:call.method]) {
        UIInterfaceOrientationMask mask = [[UIApplication sharedApplication]
            supportedInterfaceOrientationsForWindow:nil];
        UIDeviceOrientation deviceOrientation =
            [UIDevice currentDevice].orientation;
        BOOL changed = NO;
        if (deviceOrientation != UIDeviceOrientationPortrait &&
            deviceOrientation != UIDeviceOrientationPortraitUpsideDown) {
            if (mask & UIInterfaceOrientationMaskPortraitUpsideDown) {
                [[UIDevice currentDevice]
                    setValue:@(UIInterfaceOrientationPortraitUpsideDown)
                      forKey:@"orientation"];
                changed = YES;
            } else if (mask & UIInterfaceOrientationMaskPortrait) {
                [[UIDevice currentDevice]
                    setValue:@(UIInterfaceOrientationPortrait)
                      forKey:@"orientation"];
                changed = YES;
            }
        }
        [UIViewController attemptRotationToDeviceOrientation];
        result(@(changed));
    } else if ([@"setOrientationLandscape" isEqualToString:call.method]) {
        UIInterfaceOrientationMask mask = [[UIApplication sharedApplication]
            supportedInterfaceOrientationsForWindow:nil];
        UIDeviceOrientation deviceOrientation =
            [UIDevice currentDevice].orientation;
        BOOL changed = NO;
        if (deviceOrientation != UIDeviceOrientationLandscapeLeft &&
            deviceOrientation != UIDeviceOrientationLandscapeRight) {
            if (mask & UIInterfaceOrientationMaskLandscapeRight) {
                [[UIDevice currentDevice]
                    setValue:@(UIInterfaceOrientationLandscapeRight)
                      forKey:@"orientation"];
                changed = YES;
            } else if (mask & UIInterfaceOrientationMaskLandscapeLeft) {
                [[UIDevice currentDevice]
                    setValue:@(UIInterfaceOrientationLandscapeLeft)
                      forKey:@"orientation"];
                changed = YES;
            }
        }
        [UIViewController attemptRotationToDeviceOrientation];
        result(@(changed));
    } else if ([@"setOrientationAuto" isEqualToString:call.method]) {
        UIInterfaceOrientationMask mask = [[UIApplication sharedApplication]
            supportedInterfaceOrientationsForWindow:nil];

        if (mask & UIInterfaceOrientationMaskPortrait)
            [[UIDevice currentDevice] setValue:@(UIInterfaceOrientationPortrait)
                                        forKey:@"orientation"];
        [UIViewController attemptRotationToDeviceOrientation];
        result(nil);
    } else if ([@"setScreenOn" isEqualToString:call.method]) {
        bool screenOn = false;
        NSNumber* on = argsMap[@"on"];
        screenOn = on == nil ? false : on.boolValue;
        [self setScreenOn:screenOn];
        result(nil);
    } else if ([@"isScreenKeptOn" isEqualToString:call.method]) {
        bool isIdleTimerDisabled =  [[UIApplication sharedApplication] isIdleTimerDisabled];
        result(@(isIdleTimerDisabled));
    } else if ([@"brightness" isEqualToString:call.method]) {
        float brightness = [UIScreen mainScreen].brightness;
        result(@(brightness));
    } else if ([@"setBrightness" isEqualToString:call.method]) {
        float brightness = [UIScreen mainScreen].brightness;
        NSNumber* arg = argsMap[@"brightness"];
        brightness = arg == nil ? brightness : arg.floatValue;
        [[UIScreen mainScreen] setBrightness:brightness];
        result(nil);
    } else if ([@"volumeUp" isEqualToString:call.method]) {
        NSNumber *number = argsMap[@"step"];
        float step = number == nil ? _volStep : [number floatValue];
        float vol = [self getSystemVolume];
        vol += step;
        vol = [self setSystemVolume:vol];
        result(@(vol));
    } else if ([@"volumeDown" isEqualToString:call.method]) {
        NSNumber *number = argsMap[@"step"];
        float step = number == nil ? _volStep : [number floatValue];
        float vol = [self getSystemVolume];
        vol -= step;
        vol = [self setSystemVolume:vol];
        result(@(vol));
    } else if ([@"volumeMute" isEqualToString:call.method]) {
        float vol = [self setSystemVolume:0.0f];
        result(@(vol));
    } else if ([@"volumeSet" isEqualToString:call.method]) {
        NSNumber *number = argsMap[@"vol"];
        float v = number == nil ? [self getSystemVolume] : [number floatValue];
        v = [self setSystemVolume:v];
        result(@(v));
    } else if ([@"systemVolume" isEqualToString:call.method]) {
        result(@([self getSystemVolume]));
    } else if ([@"volUiMode" isEqualToString:call.method]) {
        NSNumber *number = argsMap[@"mode"];
        _volumeUIMode = [number intValue];
        [self updateVolumeVisiablity];
        result(nil);
    } else if ([@"onLoad" isEqualToString:call.method]) {
        [self initVolumeView];
        _eventListening = YES;
        result(nil);
    } else if ([@"onUnload" isEqualToString:call.method]) {
        _eventListening = NO;
        result(nil);
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)initVolumeView {
    if (_volumeView == nil) {
        _volumeView =
            [[MPVolumeView alloc] initWithFrame:CGRectMake(-100, -100, 10, 10)];
        _volumeView.hidden = NO;
    }
    if (_volumeViewSlider == nil) {
        for (UIView *view in [_volumeView subviews]) {
            if ([view.class.description isEqualToString:@"MPVolumeSlider"]) {
                _volumeViewSlider = (UISlider *)view;
                _volumeViewSlider.value = [AVAudioSession sharedInstance].outputVolume;
                break;
            }
        }
    }
    if (!_volumeInWindow) {
        UIWindow *window = UIApplication.sharedApplication.keyWindow;
        if (window != nil) {
            [window addSubview:_volumeView];
            _volumeInWindow = YES;
        }
    }
    [self updateVolumeVisiablity];
}

- (void)onPlayingChange:(int)delta {
    _playingCnt += delta;
    [self updateVolumeVisiablity];
}

- (void)onPlayableChange:(int)delta {
    _playableCnt += delta;
    [self updateVolumeVisiablity];
}

- (void)setScreenOn:(BOOL)on {
    [UIApplication sharedApplication].idleTimerDisabled = on;
}

- (float)getSystemVolume {
    [self initVolumeView];
    if (_volumeViewSlider == nil) {
        AVAudioSession *audioSession = [AVAudioSession sharedInstance];
        CGFloat currentVol = audioSession.outputVolume;
        return currentVol;
    } else {
        return _volumeViewSlider.value;
    }
}

- (float)setSystemVolume:(float)vol {
    [self initVolumeView];
    if (vol > 1.0) {
        vol = 1.0;
    } else if (vol < 0) {
        vol = 0.0;
    }
    [_volumeViewSlider setValue:vol animated:FALSE];
    vol = _volumeViewSlider.value;
    [self sendVolumeChange:vol];
    return vol;
}

- (void)updateVolumeVisiablity {
    if (_volumeView == nil || _volumeInWindow == FALSE) {
        _showOsUI = YES;
        return;
    }
    if (_volumeUIMode == alwaysShowUI) {
        _showOsUI = YES;
        _volumeView.hidden = YES;
    } else if (_volumeUIMode == neverShowUI) {
        _showOsUI = NO;
        _volumeView.hidden = NO;
    } else if (_volumeUIMode == hideUIWhenPlaying) {
        _volumeView.hidden = _playingCnt <= 0;
        _showOsUI = _volumeView.hidden;
    } else if (_volumeUIMode == hideUIWhenPlayable) {
        _volumeView.hidden = _playableCnt <= 0;
        _showOsUI = _volumeView.hidden;
    }
}

- (void)volumeChange:(NSNotification *)notifi {
    NSString *style = [notifi.userInfo
        objectForKey:@"AVSystemController_AudioCategoryNotificationParameter"];
    CGFloat value = [[notifi.userInfo
        objectForKey:@"AVSystemController_AudioVolumeNotificationParameter"]
        doubleValue];
    if ([style isEqualToString:@"Audio/Video"]) {
        [self sendVolumeChange:value];
    }
}

- (void)sendVolumeChange:(float)value {
    if (_eventListening) {
        [_eventSink success:@{
            @"event" : @"volume",
            @"sui" : @(_showOsUI),
            @"vol" : @(value)
        }];
    }
}

- (FlutterError *_Nullable)onCancelWithArguments:(id _Nullable)arguments {
    [_eventSink setDelegate:nil];
    return nil;
}

- (FlutterError *_Nullable)onListenWithArguments:(id _Nullable)arguments
                                       eventSink:
                                           (nonnull FlutterEventSink)events {
    [_eventSink setDelegate:events];
    return nil;
}

@end
